### DISCUSSION TITLE
Reuse `Decoder` allocations and settings

### DISCUSSION THREAD

**Discussion by okaneco** [image-rs/image-png#387] [Ideas]
I noticed during the overhaul of `DecodeOptions` that there doesn't seem to be a way to reuse `Decoder` allocations and settings if a user has multiple images to read. The `Decoder` allocates a `ReadDecoder`, and the `StreamingDecoder` allocates a `ZlibStream` which is where `DecodeOptions` resides.

Perhaps a `Reader::into_raw_parts` can be added that returns a `StreamingDecoder` and the three `Vec<u8>`s. A `Decoder::new_from_raw` could be built with a reset `StreamingDecoder`, and the `Reader` can reuse allocations with `Decoder::read_info_with_raw_parts(mut self, ReusedVecs) -> Result<Reader<R>> `.

I'm not 100% certain on this idea (maybe `Decoder` and `Reader` should be merged into one?), just some preliminary thoughts I wanted to share while fresh in my mind.

Struct definitions and the `ZlibStream` initialization which includes some large capacity Vecs
```rust
pub struct Decoder<R: Read> {
    read_decoder: ReadDecoder<R>,
    transform: Transformations,
    limits: Limits,
}

struct ReadDecoder<R: Read> {
    reader: BufReader<R>,
    decoder: StreamingDecoder,
    at_eof: bool,
}

ZlibStream {
    state: Box::new(Compressor::FDeflate(fdeflate::Decompressor::new())),
    started: false,
    in_buffer: Vec::with_capacity(CHUNCK_BUFFER_SIZE), // 32768
    in_pos: 0,
    out_buffer: vec![0; 2 * CHUNCK_BUFFER_SIZE],
    out_pos: 0,
    ignore_adler32: true,
}

pub struct Reader<R: Read> {
    decoder: ReadDecoder<R>,
    bpp: BytesPerPixel,
    subframe: SubframeInfo,
    fctl_read: u32,
    next_frame: SubframeIdx,
    prev: Vec<u8>,
    current: Vec<u8>,
    scan_start: usize,
    transform: Transformations,
    processed: Vec<u8>,
    limits: Limits,
}
```

**Comment by fintelia** [comment]
I think a `reset` method of some sort may be cleaner. But IMO both feel a little messy
